addons/mail/models/mail_activity.py

```diff
@@ -795,3 +796,128 @@ def activity_unlink(self, act_type_xmlids, user_id=None):
             domain = ['&'] + domain + [('user_id', '=', user_id)]
         self.env['mail.activity'].search(domain).unlink()
         return True
+
+    def _read_progress_bar(self, domain, group_by, progress_bar):
+        if not (progress_bar['field'] == 'activity_state' and self._fields[group_by].store):
+            return super()._read_progress_bar(domain, group_by, progress_bar)
+
+        # use datetime.utcnow() instead of date.today(), because
+        # it's used in _compute_state_from_date and on mocking in tests
+        today_utc = datetime.utcnow()
+        today = date(year=today_utc.year, month=today_utc.month, day=today_utc.day)
+
+        with_clause = """
+            SELECT
+                mail_activity.res_id AS res_id,
+                min(date_deadline - (%%s AT TIME ZONE res_partner.tz)::timestamp) AS days_to_deadline
+            FROM mail_activity
+            LEFT JOIN res_users ON mail_activity.user_id = res_users.id
+            LEFT JOIN res_partner ON res_users.partner_id = res_partner.id
+            WHERE mail_activity.res_model = '%(model_name)s'
+            GROUP BY mail_activity.res_id
+        """ % {
+            'model_name': self._name,
+        }
+        with_params = [today]
+
+        select_activity_state = """
+            CASE
+            WHEN mail_activity_min.days_to_deadline > interval '00:00:00' THEN 'planned'
+            WHEN mail_activity_min.days_to_deadline < interval '00:00:00' THEN 'overdue'
+            WHEN mail_activity_min.days_to_deadline = interval '00:00:00' THEN 'today'
+            ELSE null
+            END
```

Comment:
shouldn't this `CASE` be part of the `WITH` clause above? it's also very weird to use `mail_activity_min` here considering the table alias is only assigned later in the function.

Reply:
I can't be directly in that SELECT, because it depends on ``days_to_deadline`` which is computed in that select. Not sure that wrapping with-clause with another SELECT is a good idea in terms of performance.

```
odoo=# select 1 as x, x*x as y;
ERROR:  column "x" does not exist
LINE 1: select 1 as x, x*x as y;
                       ^
```


I slightly improved it by adding new variable for ``mail_activity_min``. 

My idea was to specify ``with_clause`` and ``select_activity_state`` before modified ``_read_group_raw`` to don't mix custom code with copy-pasted coded.

Reply:
> It can't be directly in that SELECT, because it depends on `days_to_deadline` which is computed in that select. 

What I meant is that you're computing `days_to_deadline` for sole purpose of using it in that `CASE`, so the WITH's SELECT could inline the CASE, as far as I can see. Using a column value several times in a SELECT clause is not more expensive, as you've already retrieved the data.

But now that we've seen the cost of the unfiltered `WITH` query on large database, it's worth investigating a direct JOIN between the model table and `mail_activity`. Let's say we have a `domain` that matches only 3 records, the current `WITH` would compute the activity_state of all records, even though we only need it for a small fraction of activities. We'd need to test it on large data sets to be sure of the best strategy.


> Not sure that wrapping with-clause with another SELECT is a good idea in terms of performance.

That's something you can verify easily ;-) The cost of executing a query is directly proportional to the work involved in retrieving the data. Wrapping a query within another query that only filters or formats data is generally negligible in terms of overhead, compared to the cost of retrieving that data in the first place.

Reply:
I summon @nseinlet

Reply:
> But now that we've seen the cost of the unfiltered `WITH` query on large database, it's worth investigating a direct JOIN between the model table and `mail_activity`. Let's say we have a `domain` that matches only 3 records, the current `WITH` would compute the activity_state of all records, even though we only need it for a small fraction of activities. We'd need to test it on large data sets to be sure of the best strategy.

I got the idea, but implementation without ``WITH`` would look too complicated in a sense that resulted code will too much differ from original ``_read_group_raw`` method.

Performance tests show good speed up. If we want further improvement, I'd prefer to do it in master

Reply:
I made additional tests, which shows that you are right -- with current implementation there is a regression on large databases and strong filter. I'll try to make the *complicated* implementation without `WITH`


demo database

BEFORE

22 0.031 0.045

AFTER

4 0.007 0.012

29 K records in crm.lead

2.

Test on My Pipeline filter (10 records)

BEFORE

22 0.012 0.032

AFTER

 4 **0.163** 0.012

Reply:
Fixed by using ```RIGHT JOIN mail_activity``` in main query and wrapping it around another query

```
            SELECT
                min(id) as id,
                SUM(__count) as __count,
                CASE
                    WHEN days_to_deadline > 0 THEN 'planned'
                    WHEN days_to_deadline < 0 THEN 'overdue'
                    WHEN days_to_deadline = 0 THEN 'today'
                    ELSE null
                END as activity_state
                %(extra_fields)s
            FROM (
              %(query)s
            ) AS x
            %(groupby)s
```